### PR TITLE
Improve Github Actions

### PR DESCRIPTION
**Pull request by erikd**
# Changelog

```yaml
- description: Improve Github Actions
# uncomment types applicable to the change:
  type:
    - maintenance    # not directly related to the code
```

# Context

Aiming to reduce the size of the GHA spec and to use the DEXV GHA helpers.

# How to trust this PR

Highlight important bits of the PR that will make the review faster. If there are commands the reviewer can run to observe the new behavior, describe them.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

**Comment by erikd** [comment]
Had a very quick look at the build speed. I found a recent `ghc-9.8` build on `Linux` that took over 26 minutes while a the same build configuration with the new caching took 23 minutes.

File: .github/workflows/haskell.yml [hunk 1]
```diff
@@ -60,49 +60,29 @@ jobs:
...
-        dist-dir: dist-newstyle
-        store-path: ${{ steps.setup-haskell.outputs.cabal-store }}
-        threads: 16
-        archive-uri: https://iohk.cache.haskellworks.io/${{ env.CABAL_CACHE_VERSION }}/${{ runner.os }}/${{ matrix.cabal }}/${{ matrix.ghc }}
-        skip: "${{ vars.CACHE_URI != '' }}"
-        enable-save: false
+        path: |
+          ${{ steps.setup-haskell.outputs.cabal-store }}
+          dist-newstyle
+        key: cache-${{ env.CABAL_CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('dependencies.txt') }}
```
Comment (smelc) [inline .github/workflows/haskell.yml:82]: > hashFiles('dependencies.txt')

I see that previously the cache wasn't indexed by the dependencies. So is there a chance this PR will reduce the occurrences of the cabal cache being corrupted (i.e. less bumps of CABAL_CACHE_VERSION required)?

**Comment by smelc** [comment]
Seems like a sane simplification :+1:

File: .github/workflows/haskell.yml [hunk 2]
```diff
@@ -60,49 +60,29 @@ jobs:
...
-      uses: action-works/cabal-cache-s3@v1
+    - name: Record dependencies
+      id: record-deps
+      run: |
+        # The tests call out to msys2 commands. We generally do not want to mix toolchains, so
+        # we are very deliberate about only adding msys64 to the path where absolutely necessary.
+        ${{ (runner.os == 'Windows' && '$env:PATH=("C:\msys64\mingw64\bin;{0}" -f $env:PATH)') || '' }}
+        cat dist-newstyle/cache/plan.json | jq -r '."install-plan"[].id' | sort | uniq > dependencies.txt
+
+    - name: Install dependencies
```
Comment (Jimbo4350) [inline .github/workflows/haskell.yml:73]: Silly question but don't we need to run the `Cache Cabal store` job first before running  `cabal build all --enable-tests --only-dependencies`?

File: .github/workflows/haskell.yml [hunk 3]
```diff
@@ -60,49 +60,29 @@ jobs:
...
+      run: cabal build all --enable-tests --only-dependencies -j --ghc-option=-j4
+
+    - name: Cache Cabal store
+      uses: actions/cache@v3
```
Comment (carbolymer) [inline .github/workflows/haskell.yml:77]: Any reason not to use `v4`?

File: .github/workflows/haskell.yml (hunk 2 above)
Comment (erikd) [reply .github/workflows/haskell.yml:73]: No, specifically want to run it after. We want to build and then store.

**Review by Jimbo4350 — APPROVED** [review]

**Comment by erikd** [comment]
As @Jimbo4350 pointed out, there was no action to restore the cabal cache. Fixing that now.

File: .github/workflows/haskell.yml (hunk 2 above)
Comment (erikd) [reply .github/workflows/haskell.yml:73]: Thanks foe making me look more closely. That made me realise that there is a cache store, but no **restore** action.

**Review by angerman — APPROVED** [review]
So the control flow is:
- dry-run build to get the proper dependency lists.
- use dependencies list as primary key for the cache.
- RESTORE the cache.
- BUILD DEPENDENCIES.
- STORE the cache.
- BUILD
- TEST
- ARTIFACTS.

Looks good to me.

**Pull request merged**